chalk-engine/src/context.rs

```diff
@@ -258,20 +239,20 @@ pub trait InferenceTable<C: Context, I: Context>:
     /// conversion -- that is, we convert the outermost goal into an
     /// `HhGoal`, but the goals contained within are left as context
     /// goals.
-    fn into_hh_goal(&mut self, goal: I::Goal) -> HhGoal<I>;
+    fn into_hh_goal(&mut self, goal: C::Goal) -> HhGoal<C>;
 
     // Used by: simplify
-    fn add_clauses(&mut self, env: &I::Environment, clauses: I::ProgramClauses) -> I::Environment;
+    fn add_clauses(&mut self, env: &C::Environment, clauses: C::ProgramClauses) -> C::Environment;
 
     /// Upcast this domain goal into a more general goal.
-    fn into_goal(&self, domain_goal: I::DomainGoal) -> I::Goal;
+    fn into_goal(&self, domain_goal: C::DomainGoal) -> C::Goal;
```

Comment:
An aside: I think it may be worth creating a first PR that **just** removes the `I: Context` type parameters and collapses to one `C: Context` everywhere, but leaves everything else the same. It'd be good to pull out and land intermediate steps.

Reply:
Well, we could probably get rid of the `WithFoo` traits and convert them to `impl FnOnce`, too, but leave everything else about them the same.

Reply:
I guess this was fairly close to your original PR :)